builder/role_image.go

```diff
@@ -50,25 +50,120 @@ type RoleImageBuilder struct {
 }
 
 // NewRoleImageBuilder creates a new RoleImageBuilder
-func NewRoleImageBuilder(repository, compiledPackagesPath, targetPath, version, fissileVersion string, ui *termui.UI) *RoleImageBuilder {
+func NewRoleImageBuilder(repository, compiledPackagesPath, targetPath, version, fissileVersion string, ui *termui.UI) (*RoleImageBuilder, error) {
+	if err := os.MkdirAll(targetPath, 0755); err != nil {
+		return nil, err
+	}
 	return &RoleImageBuilder{
 		repository:           repository,
 		compiledPackagesPath: compiledPackagesPath,
 		targetPath:           targetPath,
 		version:              version,
 		fissileVersion:       fissileVersion,
 		ui:                   ui,
+	}, nil
+}
+
+// CreatePackagesDockerBuildDir generates a Dockerfile and assets for the shared packages layer and returns a path to the dir
+func (r *RoleImageBuilder) CreatePackagesDockerBuildDir(roleManifest *model.RoleManifest, lightManifestPath, darkManifestPath string) (string, error) {
+	if len(roleManifest.Roles) == 0 {
+		return "", fmt.Errorf("No roles to build")
+	}
+
+	succeeded := false
+	buildDir, err := ioutil.TempDir(r.targetPath, "role-packages")
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		if !succeeded {
+			os.RemoveAll(buildDir)
+		}
+	}()
+	rootDir := filepath.Join(buildDir, "root")
+
+	// Generate dockerfile
+	dockerfile, err := os.Create(filepath.Join(buildDir, "Dockerfile"))
+	if err != nil {
+		return "", err
+	}
+	defer dockerfile.Close()
+	baseImageName := GetBaseImageName(r.repository, r.fissileVersion)
+	if err := r.generateDockerfile(nil, baseImageName, dockerfile); err != nil {
+		return "", err
+	}
+
+	// Generate configuration
+	r.ui.Println("Generating configuration JSON specs ...")
+	configStore := configstore.NewConfigStoreBuilder(
+		configstore.JSONProvider,
+		lightManifestPath,
+		darkManifestPath,
+		filepath.Join(rootDir, "opt/hcf/specs"),
+	)
+
+	if err := configStore.WriteBaseConfig(roleManifest); err != nil {
+		return "", fmt.Errorf("Error writing base config: %s", err.Error())
+	}
+
+	// Copy packages
+	r.ui.Println("Copying compiled packages...")
+	copiedFingerprints := make(map[string]bool)
+	for _, role := range roleManifest.Roles {
+		for _, job := range role.Jobs {
+			for _, pkg := range job.Packages {
```

Comment:
This will no longer exist after the next PR, at which point it stops mattering :grinning: 

Also, I don't think doing it in parallel would have been useful: since it's all disk-bound, doing it in parallel just means more seeks (on spinning rust) instead of building things faster.  Parallelization is way more useful for CPU-bound tasks.